service/api_logic/user_logic.py

```diff
@@ -92,34 +103,16 @@ def confirm_token(self, token: str, expiration=3600):
         return OutputUser().dump(user)
 
 
-    async def log_in(self, email_or_username: str, password: str):
-        user = self._user_dal.get_user_by_email_or_username(email_or_username, email_or_username)
-
-        if not user:
-            self._logger.warning("User does not exist")
-            raise IncorrectUsernameOrEmailError()
-
-        if not bcrypt.checkpw(password.encode('utf-8'), user.password_hash.encode('utf-8')):
-            self._logger.warning("Passwords do not match")
-            raise IncorrectPasswordError()
-
-        token = await self.__generate_auth_token(user)
-
-        return OutputLogin(email = user.email, token = token, id = user.user_id)
+    async def log_in(self, method: str, credentials: Union[InputUserLoginDTO, InputUserByGoogleDTO]):
+        if method not in self.strategies:
```

Comment:
BTW:

I'd move this strategy-selection logic somewhere closer to the strategies themselves.
It's not good to check wether the strategy is correct for log_iin and sign_up separately if we can do that in one place. Maybe let AuthContext decide which strategy has been selected. It'd become more of a factory then, but general idea remains